### PR TITLE
odb: optimize dbSet<T>::size() checks in hot loops

### PR DESCRIPTION
**Pull request by oharboe**
In dbNet::getITermCount(), GlobalRouter.cpp, TritonRoute.cpp,
Rebuffer.cc, OptMirror.cpp, and request_handler.cpp, getITerms().size()
is called inside hot loops. Because dbSet<T>::size() iterates
over a linked list, this results in O(N) traversal per call,
severely degrading performance for massive nets. This bites on
designs where an unbuffered high-fanout control net (e.g. a
synchronous reset or scan enable on an ideal network) survives
into placement and routing with 100k+ ITerms; typical designs
never notice because synthesis buffers such nets into trees.

This patch adds generic early-exit idioms hasMoreThan(N) and
hasExactly(N) to dbSet<T>, and updates the hotspots to use these
bounded checks — each walks at most N+1 elements instead of the
whole set — without touching the core memory-optimized _dbNet
struct. Unit tests cover the boundary cases (empty set, N-1, N,
N+1) on two different iterator types.

These early-exit predicates are established practice for linked
lists used as compact storage rather than an invention: LLVM's
ADT/STLExtras.h provides the same idioms for its intrusive lists
(hasNItems, hasNItemsOrMore, hasNItemsOrLess, hasSingleElement),
std::forward_list omits size() entirely so that the O(N) walk is
visible at the call site (std::distance), and the Linux kernel's
list.h names the counting walk explicitly (list_count_nodes(),
list_is_singular()).

Considered and deliberately not done, to keep this PR a single
concern:

- Making dbSet<T>::size() O(1) by caching a count: defeats the
  memory-optimized object structs; the whole point is to avoid
  needing the full count in these checks.
- Renaming or deprecating dbSet<T>::size(): it is not uniformly
  O(N) (some iterators, e.g. dbBlockItr, answer from a vector),
  and there are ~250 call sites. A rename is a separate
  discussion, not a rider on a perf fix.
- Deduplicating the identical routable-nets checks in
  TritonRoute.cpp and GlobalRouter.cpp: the duplication predates
  this PR; hoisting a shared helper into odb is scope creep here.
- Additional variants (hasAtLeast, hasFewerThan, isSingular):
  no caller needs them yet.
- noexcept / [[nodiscard]] on the new methods: no precedent in
  odb headers; style consistency wins.
- Micro-ordering of short-circuit operands among the new bounded
  checks: every operand walks at most 2 elements, all orderings
  are semantically equivalent, the difference is a few pointer
  dereferences and unmeasurable, and automated review has
  proposed both orderings across rounds. The conditions stay in
  natural reading order.
- Converting the remaining getITerms()/getBTerms().size() call
  sites in dbNet.cpp (getITermCount(), getBTermCount(), debug
  report printing): they need the actual count, which is exactly
  what size() is for.
- std::vector/std::string size() comparisons on untouched lines
  in the same files: O(1) already; out of scope.
- Harmonizing signed/unsigned count types across the codebase:
  int is the de-facto style (~150 int-typed count/limit members,
  ~200 static_cast<int>(...size()) call sites). Out of scope;
  only the two single-use constants this PR already touches are
  retyped to uint32_t, which also makes a negative threshold
  unrepresentable rather than runtime-guarded.

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

### CODE REVIEW

File: src/drt/src/TritonRoute.cpp [hunk 1]
```diff
@@ -1021,7 +1021,9 @@ int TritonRoute::main()
   initDesign();
   bool has_routable_nets = false;
   for (auto net : db_->getChip()->getBlock()->getNets()) {
-    if (net->getITerms().size() + net->getBTerms().size() > 1) {
+    if (net->getITerms().hasMoreThan(1)
+        || net->getBTerms().hasMoreThan(1)
+        || (!net->getITerms().empty() && !net->getBTerms().empty())) {
```
[the remark this answers is not in the document]
Comment (oharboe) [reply src/drt/src/TritonRoute.cpp]: Applied in the current revision.

File: src/grt/src/GlobalRouter.cpp [hunk 2]
```diff
@@ -469,7 +469,9 @@ void GlobalRouter::globalRoute(bool save_guides)
   bool has_routable_nets = false;
 
   for (auto net : db_->getChip()->getBlock()->getNets()) {
-    if (net->getITerms().size() + net->getBTerms().size() > 1) {
+    if (net->getITerms().hasMoreThan(1)
+        || net->getBTerms().hasMoreThan(1)
+        || (!net->getITerms().empty() && !net->getBTerms().empty())) {
```
[the remark this answers is not in the document]
Comment (oharboe) [reply src/grt/src/GlobalRouter.cpp]: Applied in the current revision.

File: src/web/src/request_handler.cpp [hunk 3]
```diff
@@ -2023,7 +2023,7 @@ WebSocketResponse SelectHandler::handleSchematicCone(
             }
             odb::dbNet* net = iterm->getNet();
             if (!net || seen_nets.contains(net)
-                || static_cast<int>(net->getITerms().size()) > kMaxNetFanout) {
+                || net->getITerms().hasMoreThan(kMaxNetFanout)) {
```
[the remark this answers is not in the document]
Comment (oharboe) [reply src/web/src/request_handler.cpp:2026]: Applied in the current revision.

File: src/web/src/request_handler.cpp [hunk 4]
```diff
@@ -2065,7 +2065,7 @@ WebSocketResponse SelectHandler::handleSchematicCone(
...
```
[the remark this answers is not in the document]
Comment (oharboe) [reply src/web/src/request_handler.cpp:2068]: Applied in the current revision.

File: src/odb/include/odb/dbSet.h [hunk 5]
```diff
@@ -129,10 +129,42 @@ class dbSet
...
+    iterator it = begin();
+    while (count > 0) {
+      if (it == end()) {
+        return false;
+      }
+      ++it;
+      --count;
+    }
+    return it != end();
+  }
```
[the remark this answers is not in the document]
Comment (oharboe) [reply src/odb/include/odb/dbSet.h:147]: Applied in 4afaf3059d91.

File: src/odb/include/odb/dbSet.h [hunk 6]
```diff
@@ -129,10 +129,42 @@ class dbSet
...
+    iterator it = begin();
+    while (count > 0) {
+      if (it == end()) {
+        return false;
+      }
+      ++it;
+      --count;
+    }
+    return it == end();
+  }
```
[the remark this answers is not in the document]
Comment (oharboe) [reply src/odb/include/odb/dbSet.h:177]: Applied in 4afaf3059d91.

File: src/dpl/src/OptMirror.cpp [hunk 7]
```diff
@@ -127,11 +127,10 @@ void OptimizeMirroring::findNetBoxes()
...
                   // Reducing HPWL on large nets (like clocks) is irrelevant
-                  // to mirroring criterra.
-                  // Note that getITerms().size() iteractes through the iterms
-                  // so it has to be checked once here instead of where it is
-                  // needed.
-                  || net->getITerms().size() > mirror_max_iterm_count_;
+                  // to mirroring criteria. hasMoreThan() walks the iterms
+                  // (bounded), so check once here and cache the result.
+                  || net->getITerms().hasMoreThan(
+                      static_cast<uint32_t>(mirror_max_iterm_count_));
```
[the remark this answers is not in the document]
Comment (oharboe) [reply src/dpl/src/OptMirror.cpp]: mirror_max_iterm_count_ is a compile-time constant (100); a negative value is not a reachable state, so a runtime guard would check the impossible. Fixed at the type level instead in e12401bbdd54: the constant (and kMaxNetFanout) are now uint32_t, making a negative threshold unrepresentable and removing the casts.

File: src/drt/src/TritonRoute.cpp [hunk 8]
```diff
@@ -1021,7 +1021,10 @@ int TritonRoute::main()
   initDesign();
   bool has_routable_nets = false;
   for (auto net : db_->getChip()->getBlock()->getNets()) {
-    if (net->getITerms().size() + net->getBTerms().size() > 1) {
+    auto iterms = net->getITerms();
+    auto bterms = net->getBTerms();
+    if (iterms.hasMoreThan(1) || bterms.hasMoreThan(1)
+        || (!iterms.empty() && !bterms.empty())) {
```
[the remark this answers is not in the document]
Comment (oharboe) [reply src/drt/src/TritonRoute.cpp:1027]: Declining as a class, not per suggestion: every operand here walks at most 2 elements, all orderings are semantically equivalent, and automated review has proposed both orderings across rounds (an earlier round requested the current ITerms-first order as high priority). See "Considered and deliberately not done" in the PR description.

File: src/grt/src/GlobalRouter.cpp [hunk 9]
```diff
@@ -469,7 +469,10 @@ void GlobalRouter::globalRoute(bool save_guides)
   bool has_routable_nets = false;
 
   for (auto net : db_->getChip()->getBlock()->getNets()) {
-    if (net->getITerms().size() + net->getBTerms().size() > 1) {
+    auto iterms = net->getITerms();
+    auto bterms = net->getBTerms();
+    if (iterms.hasMoreThan(1) || bterms.hasMoreThan(1)
+        || (!iterms.empty() && !bterms.empty())) {
```
[the remark this answers is not in the document]
Comment (oharboe) [reply src/grt/src/GlobalRouter.cpp:475]: Declining as a class, not per suggestion: every operand here walks at most 2 elements, all orderings are semantically equivalent, and automated review has proposed both orderings across rounds (an earlier round requested the current ITerms-first order as high priority). See "Considered and deliberately not done" in the PR description.

File: src/rsz/src/Rebuffer.cc [hunk 10]
```diff
@@ -1679,8 +1679,8 @@ static bool isPortBuffer(sta::dbNetwork* network, sta::Instance* inst)
   if (network->libertyCell(inst) && network->libertyCell(inst)->isBuffer()) {
     odb::dbInst* db_inst = network->staToDb(inst);
     for (odb::dbITerm* iterm : db_inst->getITerms()) {
-      if (iterm->getNet() && iterm->getNet()->getITerms().size() == 1
-          && !iterm->getNet()->getBTerms().empty()) {
+      odb::dbNet* net = iterm->getNet();
+      if (net && net->getITerms().hasExactly(1) && !net->getBTerms().empty()) {
```
[the remark this answers is not in the document]
Comment (oharboe) [reply src/rsz/src/Rebuffer.cc:1683]: Declining as a class, not per suggestion: every operand here walks at most 2 elements, all orderings are semantically equivalent, and automated review has proposed both orderings across rounds (an earlier round requested the current ITerms-first order as high priority). See "Considered and deliberately not done" in the PR description.

File: src/odb/src/db/dbNet.cpp [hunk 11]
```diff
@@ -1458,7 +1458,7 @@ bool dbNet::isConnected(const dbModNet* other) const
 
 bool dbNet::isConnectedByAbutment()
 {
-  if (getITermCount() > 2 || getBTermCount() > 0) {
+  if (getITerms().hasMoreThan(2) || !getBTerms().empty()) {
```
[the remark this answers is not in the document]
Comment (oharboe) [reply src/odb/src/db/dbNet.cpp:1461]: Declining as a class, not per suggestion: every operand here walks at most 2 elements, all orderings are semantically equivalent, and automated review has proposed both orderings across rounds (an earlier round requested the current ITerms-first order as high priority). See "Considered and deliberately not done" in the PR description.

**Comment by oharboe** [comment]
@maliberty Getting off the gemini nit-picking hamsterwheel, it isn't converging....

**Review by maliberty — APPROVED** [review]

**Comment by maliberty** [comment]
Did you measure a benefit from this?

**Comment by oharboe** [comment]
> Did you measure a benefit from this?

Yes, I think on a design like MegaBoom, this is ca. 35% of the running time in gpl because there are an enormous fanout of synchronous reset. At least that's the best theory that agy and I could come up with :-) Now that this 35% is fixed, what was 20% in omp should be something like 30% of the running time. I'm inclined to believe there's some substantial amount of AI induced fantasies here, but the reason I latched on to this was because ::size() was top of `perf top`.

I suppose this is unobservable on the designs that are regularly used for testing, or it would have been fixed already...
